{% extends 'base.html' %}

{% block content %}

<h2>Get Started</h2>

<ul>
  <li class="h5">
    View <a target="_blank" href="https://crispycrunch.czbiohub.org/main/primer-selection/91/experiment-summary/">demo experiment</a>
    and <a target="_blank" href="https://crispycrunch.czbiohub.org/main/analysis/11/results/?url_auth_token=AAAAD6DADtu-27EV2aBL2x6BrSY%3AsYYjdE_ZJHYKAFvy7XjberSIR1c">demo analysis</a>
  </li>
  <li class="h5">Read <a href="/howto/">how to guide</a></li>
  <li class="h5">Create an <a href="/main/signup/">account</a></li>
  <li class="h5">Design a <a href="/main/experiment/">new experiment</a></li>
  <li class="h5">Start a <a href="/main/analysis/">new analysis</a></li>
  <li class="h5">View your <a href="/main/">current experiments and analyses</a></li>
</ul>

<h2>Introduction</h2>

{# TODO (gdingle): link to addgene post when ready or own documentation #}

<p>CrispyCrunch is a tool for defining <strong>batches</strong> of <a target="_blank" href="https://en.wikipedia.org/wiki/CRISPR">CRISPR</a> edits and analyzing their basic effects. An edit can be either a plain CRISPR/Cas9 cut––a "<strong>knock-out</strong>"––or a cut plus insertion through <a target="_blank" href="https://en.wikipedia.org/wiki/Homology_directed_repair">HDR</a>––a "<strong>knock-in</strong>". It was initially built for the <a target="_blank" href="https://www.czbiohub.org">CZ Biohub</a> project by <a target="_blank" href="https://www.czbiohub.org/people3/cz-biohub-staff#l-start">Manuel Leonetti</a> to tag all <a target="_blank" href="https://en.wikipedia.org/wiki/Human_genome#Coding_sequences_(protein-coding_genes)">22,000 protein coding genes</a> in the human genome with <a target="_blank" href="https://en.wikipedia.org/wiki/Green_fluorescent_protein">GFP</a>.</p>

<p>
  <figure class="figure shadow p-3 mb-3 bg-white rounded ml-5">
    <img src="/static/crispycrunch-example.png" width="1024"
      class="figure-img img-fluid rounded" />
    <figcaption class="figure-caption text-center">
      Example batch CRISPR+HDR experiment from CrispyCrunch
    </figcaption>
  </figure>
</p>

<p>CrispyCrunch uses the popular tool <a target="_blank" href="http://crispor.tefor.net/"><strong>Crispor</strong></a> for finding potential <strong>guide sequences</strong>––sgRNAs––and for designing <strong>primers</strong> for <a target="_blank" href="https://en.wikipedia.org/wiki/Polymerase_chain_reaction">PCR</a> amplification. (Crispor in turn uses the ubiquitous <a target="_blank" href="https://primer3plus.com/">primer3</a> for primer design.) Click here to see <a target="_blank" href="http://ec2-34-219-237-20.us-west-2.compute.amazonaws.com/crispor.py?batchId=gYvicTzp9e5VPFC9YwLR">an example set of guide sequences</a> from Crispor.</p>

<p>For analysis, CrispyCrunch uses another popular tool, <a target="_blank" href="http://crispresso.pinellolab.partners.org/"><strong>Crispresso</strong></a>. The analysis quantifies the <strong>efficiency</strong> of knock-in or knock-out edits by <strong>repair outcome</strong>––<a target="_blank" href="https://en.wikipedia.org/wiki/Non-homologous_end_joining">NHEJ</a> or <a target="_blank" href="https://en.wikipedia.org/wiki/Homology_directed_repair">HDR</a>. Click here to see <a target="_blank" href="http://crispresso.pinellolab.partners.org/view_report/dlJzEa">an example analysis report</a> from Crispresso.</p>

<p>
  {# TODO (gdingle): background gradient #}
  <figure class="figure shadow p-3 mb-3 bg-white rounded ml-5">
    <img src="/static/crispor-example.png" width="640"
      class="figure-img img-fluid rounded" />
    <figcaption class="figure-caption text-center">
      Guide sequences from Crispor
    </figcaption>
  </figure>

  <figure class="figure shadow p-3 mb-3 bg-white rounded ml-5">
    <img src="/static/crispresso-example.png" width="360"
      class="figure-img img-fluid rounded" />
    <figcaption class="figure-caption text-center">
      Edit efficiency from Crispresso
    </figcaption>
  </figure>
</p>

<h2>Key Features</h2>

<p>The advantages of using CrispyCrunch over the underlying tools are:</p>
<ul style="columns: 2;">
  <li>Design <strong>multiple CRISPR edits</strong> as unit and analyze them as a unit. Save time and avoid mistakes by not doing your own record keeping.</li>
  <li><strong>Automatic quality control</strong> of CRISPR edits. Just point CrispyCrunch to the FastQ output files of your <a target="_blank" href="https://en.wikipedia.org/wiki/DNA_sequencer">DNA sequencer</a>.</li>
  <li>Benefit from <strong>sophisticated HDR design rules</strong> which consider cut-to-insert distance, re-cutting, synonymous mutations, optimal primer length, intron/exon junctions, UTR boundaries and more.</li>
  <li><strong>Pre-filled order forms</strong> for reagent vendors.</li>
</ul>

<h2>Related works</h2>

{# TODO (gdingle): bootstrap style #}
<ul style="columns: 2;">
  <li><a href="http://crispor.tefor.net/">Crispor</a></li>
  <li><a href="http://crispresso.pinellolab.partners.org/">Crispresso</a></li>
  <li><a href="http://tagin.stembio.org/">TagIn</a></li>
  <li><a href="https://bioconductor.org/packages/release/bioc/html/CrispRVariants.html">CrispRVariants
</a></li>
  <li><a href="https://bioconductor.org/packages/release/bioc/html/amplican.html">AmpliCan
</a></li>
</ul>

<h2>Acknowledgements</h2>

<ul style="columns: 2;">
  <li>Aaron McGeever</li>
  <li>Ben Southgate</li>
  <li>David Dynerman</li>
  <li>Giana Cirolia</li>
  <li>Jason Li</li>
  <li>Joe Hiatt</li>
  <li>Kendell Clement</li>
  <li>Luca Pinello</li>
  <li>Manuel Leonetti</li>
  <li>Martin Kampmann</li>
  <li>Max Haeussler</li>
  <li>Ryan Leenay</li>
  <li>Steve Pollard</li>
  <li>Theo Roth</li>
  <li><b>Many other people at CZI and the Biohub</b></li>
</ul>

<h2>Feedback</h2>

{# TODO (gdingle): link to github #}
<p>Questions and comments are welcome! Please email <a href="mailto:{{ settings.ADMIN_EMAIL }}?subject=CrispyCrunch">{{ settings.ADMIN_EMAIL }}</a>.</p>

{% endblock %}
